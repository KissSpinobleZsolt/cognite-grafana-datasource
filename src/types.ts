// eslint-disable-next-line max-classes-per-file
import {
  DataQueryRequest,
  DataQuery as DataQueryUI,
  TimeSeries,
  DataQuery,
  DataSourceJsonData,
} from '@grafana/data';
import { TimeSeriesResponseItem, Datapoints, Items, IdEither, Limit } from './cdf/types';

export interface CogniteQuery extends DataQuery {
  target: number | ''; // Timeseries internal Id:
  aggregation: string;
  granularity: string;
  error: string;
  label: string;
  tab: Tab;
  assetQuery: AssetQuery;
  expr: string;
  warning: string;
}

export enum Tab {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Custom = 'Custom',
  Template = 'Template',
}

export const defaultAssetQuery: AssetQuery = {
  includeSubtrees: false,
  target: '',
};

// TODO: Investigate if "type" property is required, it is currently not defined in MyQuery
// These defaults are extracted from the old queryCtrl.ts
export const defaultQuery: Partial<CogniteQuery> = {
  target: '',
  // type: 'timeserie',
  aggregation: 'average',
  granularity: '',
  label: '',
  tab: Tab.Timeseries,
  expr: '',
  assetQuery: {
    target: '',
    includeSubtrees: false,
  },
};

/**
 * These are options configured for each DataSource instance
 */

export interface CogniteDataSourceOptions extends DataSourceJsonData {
  cogniteApiUrl?: string;
  authType: string;
  defaultRegion: string;
  cogniteProject: string;
}

export interface MySecureJsonData {
  apiKey?: string;
}

/**
 * Value that is used in the backend, but never sent over HTTP to the frontend
 */
export function isError(maybeError: DataQueryError | any): maybeError is DataQueryError {
  return (<DataQueryError>maybeError).error !== undefined;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 * @param name â€“ event name
 */
export const eventFactory = <T = undefined>(name: string): AppEvent<T> => {
  return { name };
};

export type QueryResponse = DataResponse<TimeSeries[]>;

export interface MetricDescription {
  readonly text: string;
  readonly value: number | string;
}

export type MetricFindQueryResponse = MetricDescription[];

export enum ParseType {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Event = 'Event',
}

export interface AssetQuery {
  target: string;
  includeSubtrees: boolean;
  old?: AssetQuery;
  timeseries?: TimeSeriesResponseItem[];
  func?: string;
  templatedTarget?: string;
}

export interface InputQueryTarget extends DataQueryUI {
  target: number | '';
  aggregation: string;
  granularity: string;
  error: string;
  label: string;
  tab: Tab;
  assetQuery: AssetQuery;
  expr: string;
  warning: string;
  templateQuery: TemplateQuery;
}

export interface QueryTarget extends InputQueryTarget {
  target: number;
}

export type QueryFormat = 'json';

export type QueryOptions = DataQueryRequest<InputQueryTarget>;

export type Tuple<T> = [T, T];

/**
 * Comes from grafana, could be imported in future releases hopefully
 */
export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export interface DataSourceRequestOptions {
  url: string;
  method: HttpMethod;
  retry?: number;
  requestId?: string;
  headers?: { [s: string]: string };
  silent?: boolean;
  data?: any;
}

export type SuccessResponse = {
  metadata: ResponseMetadata;
  result: DataQueryRequestResponse;
};

export type FailResponse = {
  metadata: ResponseMetadata;
  error: any;
};

export abstract class Result<T, E> {
  abstract isOk: boolean;
  abstract isErr: boolean;
}
export class Ok<T, E> implements Result<T, E> {
  value: T;

  constructor(v: T) {
    this.value = v;
  }

  isOk: boolean = true;
  isErr: boolean = false;
}

export class Err<T, E> implements Result<T, E> {
  error: E;

  constructor(error: E) {
    this.error = error;
  }

  isOk: boolean = false;
  isErr: boolean = true;
}

export type Responses<T, U> = {
  succeded: T[];
  failed: U[];
};

export interface DataQueryRequestResponse extends DataResponse<Datapoints> {
  config: {
    data: {
      aggregates: string;
      limit: number;
    };
  };
}

export interface RequestParams<DataType = any> {
  path: string;
  data: DataType;
  method: HttpMethod;
  params?: { [s: string]: any };
  requestId?: string;
  cacheTime?: string;
}

export type DataQueryError = {
  error: {
    cancelled?: boolean;
    data: {
      error?: {
        message: string;
        notFound?: string[];
      };
    };
    status: number;
  };
};

export type QueriesData = {
  items: DataQueryRequestItem[];
  target: QueryTarget;
}[];

export type ResponseMetadata = { labels: string[]; target: QueryTarget };

export interface DataQueryAlias {
  alias: string;
  id: number;
  aggregate?: string;
  granularity?: string;
}

export type DataQueryRequestItem = {
  expression?: string;
  start?: string | number;
  end?: string | number;
  limit?: number;
  granularity?: string;
  aggregates?: string[];
  id?: number;
};

export type Aggregates = Pick<CDFDataQueryRequest, 'aggregates'>;
export type Granularity = Pick<CDFDataQueryRequest, 'granularity'>;

export interface CDFDataQueryRequest {
  items: DataQueryRequestItem[];
  start?: string | number;
  end?: string | number;
  limit?: number;
  aggregates?: string[];
  granularity?: string;
}

export interface CogniteAnnotationQuery extends DataQuery {
  query?: string;
  error?: string;
}

export interface AnnotationSearchQuery {
  description: string;
  type: string;
  subtype: string;
  minStartTime: number;
  maxStartTime: number;
  minEndTime: number;
  maxEndTime: number;
  minCreatedTime: number;
  maxCreatedTime: number;
  minLastUpdatedTime: number;
  maxLastUpdatedTime: number;
  // format is {"k1": "v1", "k2": "v2"}
  metadata: string;
  assetIds: number[];
  assetSubtrees: number[];
  sort: 'startTime' | 'endTime' | 'createdTime' | 'lastUpdatedTime';
  dir: 'asc' | 'desc';
  limit: number;
  offset: 0;
}

export interface DataResponse<T> {
  data: T;
}

export type CursorResponse<T> = DataResponse<Items<T> & { nextCursor?: string }>;

export type Response<T = any> = DataResponse<{
  items: T[];
}>;

export type TimeseriesFilterQuery = {
  filter?: {
    description?: string;
    assetSubtreeIds?: IdEither[];
    assetIds?: string[];
    [s: string]: any; // so we auto-support next features
  };
  cursor?: string;
} & Limit;

export interface VariableQueryData {
  query: string;
  error?: string;
}

export interface VariableQueryProps {
  query: string;
  onChange: (query: VariableQueryData, description: string) => void;
  datasource: any;
  templateSrv: any;
}

export interface QueryRequestError {
  refId: string;
  error: string;
}

export interface QueryDatapointsWarning {
  refId: string;
  warning: string;
}

export interface TemplateQuery extends DataQuery {
  domain: string;
  domainVersion: number;
  queryText: string;
  dataPath: string;
  dataPointsPath: string;
  groupBy: string;
  aliasBy: string;
  annotationTitle: string;
  annotationText: string;
  annotationTags: string;
  constant: number;
}

export const defaultTemplateQuery: Partial<TemplateQuery> = {
  domain: undefined,
  domainVersion: undefined,
  queryText: `query {
      wellList {
        name,
        pressure {
          datapoints(start: $__from, end: $__to, limit: 50) {
            timestamp,
            value
          }
        }
      }
  }`,
  dataPath: 'data',
  dataPointsPath: '',
  groupBy: '',
  aliasBy: '',
  annotationTitle: '',
  annotationText: '',
  annotationTags: '',
  constant: 6.5,
};
