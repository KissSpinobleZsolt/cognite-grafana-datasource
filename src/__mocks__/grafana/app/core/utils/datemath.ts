// for tests, just pass in the millisecond timestamps as strings
export function parse(text: any, roundUp?: any): any {
  return Number(text);
}
