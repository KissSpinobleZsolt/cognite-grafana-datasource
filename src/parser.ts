import {
  DataQueryRequestItem,
  ParseType,
  QueryOptions,
  FilterOptions,
  FilterType,
  TimeSeriesResponseItem,
  DataQueryAlias,
  QueryTarget,
} from './types';
import Utils from './utils';
import _ from 'lodash';
import { TemplateSrv } from 'grafana/app/features/templating/template_srv';

export const parseExpression = (
  expr: string,
  options: QueryOptions,
  timeseries: TimeSeriesResponseItem[],
  templateSrv: TemplateSrv,
  target: QueryTarget
): DataQueryRequestItem[] => {
  const trimmedExpr = expr.trim();
  // first check if it is just a simple `timeseries{}` or `timeseries{}[]`
  if (isSimpleTimeseriesExpression(trimmedExpr)) {
    const filterOptions = getAndApplyFilterOptions(trimmedExpr, templateSrv, options, timeseries);
    target.aggregation = filterOptions.aggregation;
    target.granularity = filterOptions.granularity;
    return timeseries.filter(ts => ts.selected).map(ts => ({ name: ts.name }));
  }

  // add special function name if only sum,avg,etc?
  const exprWithSpecialFunctions = parseSpecialFunctions(
    trimmedExpr,
    options,
    timeseries,
    templateSrv
  );

  return createDataQueryRequestItems(
    exprWithSpecialFunctions,
    options,
    timeseries,
    templateSrv,
    ''
  );
};

const parseSpecialFunctions = (
  expr: string,
  options: QueryOptions,
  timeseries: TimeSeriesResponseItem[],
  templateSrv: TemplateSrv
) => {
  let newExpr = expr;
  // look for sum() - case-insensitive
  const sumRegex = /sum\(.*?\)/gi;
  let sumRegexMatches = newExpr.match(sumRegex);
  if (sumRegexMatches) {
    // remove duplicates
    sumRegexMatches = sumRegexMatches.filter((match, i, a) => i === a.indexOf(match));
    for (const match of sumRegexMatches) {
      const timeseriesString = match.slice(4, -1);
      const filterOptions = getAndApplyFilterOptions(
        timeseriesString,
        templateSrv,
        options,
        timeseries
      );
      const selectedTs = timeseries.filter(ts => ts.selected);
      const sumString = `([${selectedTs
        .map(ts => getTempAliasString(ts, filterOptions))
        .join('] + [')}])`;
      newExpr = newExpr.replace(match, sumString);
    }
  }

  // look for max(), min(), or avg()
  const funcRegex = /(max|min|avg)\(.*?\)/gi;
  let funcRegexMatches = newExpr.match(funcRegex);
  if (funcRegexMatches) {
    // remove duplicates
    funcRegexMatches = funcRegexMatches.filter((match, i, a) => i === a.indexOf(match));
    for (const match of funcRegexMatches) {
      const timeseriesString = match.slice(4, -1);
      const filterOptions = getAndApplyFilterOptions(
        timeseriesString,
        templateSrv,
        options,
        timeseries
      );
      const selectedTs = timeseries.filter(ts => ts.selected);
      let funcString = '';
      if (selectedTs.length <= 1) {
        funcString = selectedTs.map(ts => `[${getTempAliasString(ts, filterOptions)}]`).join('');
      } else {
        funcString = `${match.slice(0, 3)}([${selectedTs
          .map(ts => getTempAliasString(ts, filterOptions))
          .join('], [')}])`;
      }
      newExpr = newExpr.replace(match, funcString);
    }
  }

  return newExpr;
};

// recursively calls itself to create an array of DataQueryRequestItems
const createDataQueryRequestItems = (
  expr: string,
  options: QueryOptions,
  timeseries: TimeSeriesResponseItem[],
  templateSrv: TemplateSrv,
  name: string
): DataQueryRequestItem[] => {
  let dataItems: DataQueryRequestItem[] = [];
  // match timeseries{}[] or timeseries{}
  const timeseriesRegex = /timeseries\{.*?\}(?:\[.*?\])?/;
  const timeseriesMatch = expr.match(timeseriesRegex);
  if (!timeseriesMatch) {
    dataItems.push({
      name,
      function: expr,
    });
  } else {
    const filterOptions = getAndApplyFilterOptions(
      timeseriesMatch[0],
      templateSrv,
      options,
      timeseries
    );
    const selectedTs = timeseries.filter(ts => ts.selected);

    for (const ts of selectedTs) {
      const replaceString = `[${getTempAliasString(ts, filterOptions)}]`;
      dataItems = dataItems.concat(
        createDataQueryRequestItems(
          expr.replace(timeseriesMatch[0], replaceString),
          options,
          timeseries,
          templateSrv,
          name || ts.name
        )
      );
    }
  }

  return dataItems.map(item => {
    updateAliases(item, options);
    return item;
  });
};

const getAndApplyFilterOptions = (
  timeseriesString: string,
  templateSrv: TemplateSrv,
  options: QueryOptions,
  timeseries: TimeSeriesResponseItem[]
) => {
  const filterOptions = parse(timeseriesString, ParseType.Timeseries, templateSrv, options);
  if (filterOptions.error) throw filterOptions.error;
  Utils.applyFilters(filterOptions.filters, timeseries);
  return filterOptions;
};

// puts in format: 'ID' or 'ID,aggr' or 'ID,aggr,gran'
const getTempAliasString = (timeseries: TimeSeriesResponseItem, filterOptions: FilterOptions) => {
  return `${timeseries.id}${filterOptions.aggregation ? `,${filterOptions.aggregation}` : ''}${
    filterOptions.granularity ? `,${filterOptions.granularity}` : ''
  }`;
};

const isSimpleTimeseriesExpression = (expr: string) => {
  const timeseriesRegex = /^timeseries\{.*?\}(?:\[.*?\])?/;
  const timeseriesMatch = expr.match(timeseriesRegex);
  if (timeseriesMatch && timeseriesMatch[0] === expr) return true;
  return false;
};

// take in a dataqueryrequestitem and replace all [ID,agg] or [ID,agg,gran] with aliases
const updateAliases = (queryItem: DataQueryRequestItem, options: QueryOptions) => {
  const regexSearch = /\[.*?\]/g;
  const regexMatches = queryItem.function.match(regexSearch);
  if (!queryItem.aliases) queryItem.aliases = [];
  if (regexMatches) {
    for (const match of regexMatches) {
      // format is id, aggregation, granularity
      const aliasParts = match
        .substr(1, match.length - 2)
        .split(',')
        .filter(string => string.length)
        .map(x => _.trim(x, ' \'"'));
      // if we only get [ID] or [ALIAS], then there is no need to make an alias
      if (aliasParts.length === 1) continue;
      const alias: DataQueryAlias = {
        alias: `alias${aliasParts.join('_')}`,
        id: Number(aliasParts[0]),
      };
      alias.aggregate = aliasParts[1];
      alias.granularity = aliasParts[2] || Utils.intervalToGranularity(options.intervalMs);
      queryItem.function = queryItem.function.replace(match, `[${alias.alias}]`);
      if (queryItem.aliases.find(x => x.alias === alias.alias)) continue;
      queryItem.aliases.push(alias);
    }
  }
};

export const parse = (
  customQuery: string,
  type: ParseType,
  templateSrv: TemplateSrv,
  options?: QueryOptions
): FilterOptions => {
  let query = customQuery;
  if (type === ParseType.Timeseries || type === ParseType.Event) {
    // replace variables with their values
    if (options) {
      query = templateSrv.replace(query, options.scopedVars);
    } else {
      for (const templateVariable of templateSrv.variables) {
        query = query.replace(`[[${templateVariable.name}]]`, templateVariable.current.value);
        query = query.replace(`$${templateVariable.name}`, templateVariable.current.value);
      }
    }
  }

  const filtersOptions = {
    filters: [],
    granularity: '',
    aggregation: '',
    error: '',
  };

  // Format: timeseries{ options }
  //     or  timeseries{ options }[aggregation, granularity]
  // regex pulls out the options string, as well as the aggre/gran string (if it exists)
  const timeseriesRegex = /^timeseries\{(.*)\}(?:\[(.*)\])?$/;
  const timeseriesMatch = query.match(timeseriesRegex);
  const assetRegex = /^(?:asset|event)\{(.*)\}$/;
  const assetMatch = query.match(assetRegex);
  const filterRegex = /^filter\{(.*)\}$/;
  const filterMatch = query.match(filterRegex);

  let splitfilters: string[];
  if (timeseriesMatch) {
    // regex finds commas that are not followed by a closed bracket
    splitfilters = Utils.splitFilters(timeseriesMatch[1], filtersOptions, false);
  } else if (assetMatch) {
    splitfilters = Utils.splitFilters(assetMatch[1], filtersOptions, true);
  } else if (filterMatch) {
    splitfilters = Utils.splitFilters(filterMatch[1], filtersOptions, false);
  } else {
    filtersOptions.error = `ERROR: Unable to parse expression ${query}`;
  }

  if (filtersOptions.error) {
    return filtersOptions;
  }

  for (let f of splitfilters) {
    f = _.trim(f, ' ');
    if (f === '') continue;
    const filter: any = {};
    let i: number;
    if ((i = f.indexOf(FilterType.RegexEquals)) > -1) {
      filter.property = _.trim(f.substr(0, i), ' \'"');
      filter.value = _.trim(f.substr(i + 2), ' \'"');
      filter.type = FilterType.RegexEquals;
    } else if ((i = f.indexOf(FilterType.RegexNotEquals)) > -1) {
      filter.property = _.trim(f.substr(0, i), ' \'"');
      filter.value = _.trim(f.substr(i + 2), ' \'"');
      filter.type = FilterType.RegexNotEquals;
    } else if ((i = f.indexOf(FilterType.NotEquals)) > -1) {
      filter.property = _.trim(f.substr(0, i), ' \'"');
      filter.value = _.trim(f.substr(i + 2), ' \'"');
      filter.type = FilterType.NotEquals;
    } else if ((i = f.indexOf(FilterType.Equals)) > -1) {
      filter.property = _.trim(f.substr(0, i), ' \'"');
      filter.value = _.trim(f.substr(i + 1), ' \'"');
      filter.type = FilterType.Equals;
    } else {
      console.error(`Error parsing ${f}`);
    }
    filtersOptions.filters.push(filter);
  }

  if (timeseriesMatch) {
    const aggregation = timeseriesMatch[2];
    if (aggregation) {
      const splitAggregation = aggregation.split(',');
      filtersOptions.aggregation = _.trim(splitAggregation[0], ' \'"').toLowerCase();
      filtersOptions.granularity =
        splitAggregation.length > 1 ? _.trim(splitAggregation[1], ' \'"') : '';
    }
  }

  return filtersOptions;
};
