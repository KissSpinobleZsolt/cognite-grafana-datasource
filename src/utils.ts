import _ from 'lodash';

export default class Utils {
  // Converts an object to a query string, ignores properties with undefined/null values
  // TODO: maybe clean this up a bit, might break easily
  static getQueryString(obj: any) {
    return _.reduce(
      obj,
      (result: string, val: any, key: string) => {
        return _.isNil(val)
          ? result
          : _.isArray(val)
          ? `${result + [key, val].map(encodeURIComponent).join('=[')}]&`
          : `${result + [key, val].map(encodeURIComponent).join('=')}&` + '';
      },
      ''
    ).slice(0, -1);
  }

  static getDatasourceValueString(aggregation: string): string {
    const mapping = {
      undefined: 'value',
      none: 'value',
      avg: 'average',
      int: 'interpolation',
      stepinterpolation: 'stepInterpolation',
      step: 'stepInterpolation',
      continuousvariance: 'continousVariance', // spelling mistake is intended - will have to change in 0.6
      continuousVariance: 'continousVariance',
      cv: 'continousVariance',
      discretevariance: 'discreteVariance',
      dv: 'discreteVariance',
      totalvariation: 'totalVariation',
      tv: 'totalVariation',
    };
    return mapping[aggregation] || aggregation;
  }
}
