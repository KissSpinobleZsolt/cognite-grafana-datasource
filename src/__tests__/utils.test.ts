import { applyFilters, flatten, isRFC3339_ISO6801 } from '../utils';
import { FilterType, ParsedFilter } from '../parser/types';
import { isArray } from 'lodash';

const { NotEquals, RegexNotEquals, RegexEquals } = FilterType;

describe('Utils', () => {
  describe('Filtering', () => {
    const assets = [
      { id: 123, name: 'asset 1', description: 'test asset 1', metadata: { key1: 'value1' } },
      { id: 456, name: 'asset 2', description: 'test asset 2', metadata: { key1: 'value2' } },
      {
        id: 789,
        name: 'asset 3',
        description: 'test asset 3',
        metadata: { key1: 'value3', key2: 'value2' },
        flag: true,
      },
      { id: 999, name: 'foo', description: 'bar', metadata: { key1: 'value1' }, flag: false },
    ];
    const rootFilter = { path: 'name', filter: NotEquals, value: 'foo' } as ParsedFilter;
    const idFilter = { path: 'id', filter: NotEquals, value: 123 } as ParsedFilter;
    const boolFilter = { path: 'flag', filter: NotEquals, value: true } as ParsedFilter;
    const notEndWith = {
      path: 'metadata.key1',
      filter: RegexNotEquals,
      value: '.*1',
    } as ParsedFilter;
    const caseSensitive = {
      path: 'metadata.key1',
      filter: RegexEquals,
      value: 'VAL.*',
    } as ParsedFilter;
    const noFilterField = {
      path: 'metadata.key2',
      filter: RegexNotEquals,
      value: '.*3',
    } as ParsedFilter;
    const filters = [notEndWith, rootFilter];
    it('should be case sensitive', () => {
      expect(applyFilters(assets, [caseSensitive]).length).toEqual(0);
    });
    it('should filter root props', () => {
      expect(applyFilters(assets, [rootFilter]).length).toEqual(3);
    });
    it('should filter numbers and booleans', () => {
      expect(applyFilters(assets, [idFilter, boolFilter]).length).toEqual(2);
    });
    it('should not drop all results which miss filter key', () => {
      expect(applyFilters(assets, [noFilterField]).length).toEqual(4);
    });
    it('should follow "and" logic for multiple filter', () => {
      expect(applyFilters(assets, filters).length).toEqual(2);
    });
  });

  describe('Flatten', () => {
    const obj = {
      undefined,
      null: null,
      string: 'hello',
      number: 123,
      float: 123.4,
      array: [1, 2, 3],
      nested: {
        undefined,
        null: null,
        string: 'hello',
        number: 123,
        float: 123.4,
      },
    };

    const flattenObj = {
      undefined,
      null: null,
      string: 'hello',
      number: 123,
      float: 123.4,
      array: [1, 2, 3],
      'nested.string': 'hello',
      'nested.number': 123,
      'nested.float': 123.4,
      'nested.null': null,
      'nested.undefined': undefined,
    };
    it('should flatten nested objects to dot notation', () => {
      expect(flatten(obj)).toEqual(flattenObj);
    });
    it('should not flatten arrays', () => {
      expect(isArray(flatten(obj).array)).toBeTruthy();
    });
  });

  describe('RFC3339 and ISO8601 valid string test', () => {
    it('should return false on bad string', () => {
      expect(isRFC3339_ISO6801('I am not a date but a string')).toBe(false);
    });
    it('should return false on a short numeric string', () => {});
    expect(isRFC3339_ISO6801('1234')).toBe(false);
    it('should return false on short numebrs', () => {
      expect(isRFC3339_ISO6801(8)).toBe(false);
    });
    it('should return false on null', () => {
      expect(isRFC3339_ISO6801(null)).toBe(false);
    });
    it('should return true on valid formatted date string with ms', () => {
      expect(isRFC3339_ISO6801('2020-06-01T00:00:00.000Z')).toBe(true);
    });
    it('should  return true on valid formatted date string without ms', () => {
      expect(isRFC3339_ISO6801('2020-06-01T00:00:00Z')).toBe(true);
    });
    it('should return false on true', () => {
      expect(isRFC3339_ISO6801(true)).toBe(false);
    });
    it('should return false on zero', () => {
      expect(isRFC3339_ISO6801(0)).toBe(false);
    });
    it('should return false on decimal', () => {
      expect(isRFC3339_ISO6801(0.111111)).toBe(false);
    });
  });
});
