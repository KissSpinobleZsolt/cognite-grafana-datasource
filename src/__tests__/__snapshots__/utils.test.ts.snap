// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils generateNodesAndEdges should match snapshot 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "title",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "nodes",
    "refId": undefined,
  },
  Object {
    "fields": Array [
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "id",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "source",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "target",
        "type": "string",
        "values": Array [],
      },
      Object {
        "config": Object {},
        "labels": undefined,
        "name": "mainStat",
        "type": "string",
        "values": Array [],
      },
    ],
    "meta": Object {
      "preferredVisualisationType": "nodeGraph",
    },
    "name": "edges",
    "refId": undefined,
  },
]
`;
